### ISSUE TITLE
Ivy is leaking memory

### ISSUE THREAD

**Issue by Achilles1515** [angular/angular#35164]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/core@9.0.0-rc.14 and earlier versions as well.

### Is this a regression?

 Yes, the previous version in which this bug was not present was: View Engine

### Description
Ivy is leaking memory by holding onto component references in some manner.

## 🔬 Minimal Reproduction

[https://ng-run.com/edit/tvqV5P78BchpccNmW4XU?open=app%2Fapp.component.ts&aot=true](https://ng-run.com/edit/tvqV5P78BchpccNmW4XU?open=app%2Fapp.component.ts&aot=true)

All this application does is create a dynamic component (TabComponent) and then immediately destroy it, within the AppComponent ngOnInit method.

In Chrome, take a memory heap snapshot of the 'ng-run-preview.firebaseapp.com' instance.
Search for `tabcomponent`:
![image](https://user-images.githubusercontent.com/23387905/73848921-05525f00-47f7-11ea-9257-7009a8f4e2a0.png)
...the destroyed component is still in memory.

On the left-hand sidebar of ng-run, click "Base Project" and then turn "Ivy AOT" off.

After the app reloads, take another heap snapshot and perform the same search.
![image](https://user-images.githubusercontent.com/23387905/73849167-772aa880-47f7-11ea-9231-52f7fd1f5df1.png)
...the destroyed component is gone from memory, which is expected.

Note: This issue might be related to #35148

**Comment by Achilles1515** [comment]
@christianliebel, as I referenced your issue.

**Comment by pkozlowski-opensource** [comment]
I can still reproduce it with 9.0.2: https://ng-run.com/edit/4XXTKV7ku6JYb3zkR495

**Comment by Achilles1515** [comment]
> I can still reproduce it with 9.0.2: https://ng-run.com/edit/4XXTKV7ku6JYb3zkR495

With your demo, I see the following in my heap dump:
![image](https://user-images.githubusercontent.com/23387905/75164200-6b434f80-56ee-11ea-870b-771c2a146014.png)

...which is different than my screenshot from 9.0 which kept the actual `TabComponent` instance in memory. Should `LComponentView_TabComponent` still be in memory? I don't know the details of the full purpose of this data structure.

**Comment by pkozlowski-opensource** [comment]
@Achilles1515 interesting, interesting, this is _not_ what I'm seeing:

![Screen Shot 2020-02-24 at 16 22 15](https://user-images.githubusercontent.com/973550/75164971-ef62fa80-5721-11ea-8c77-ff9984135354.png)

**Comment by Achilles1515** [comment]
> @Achilles1515 interesting, interesting, this is _not_ what I'm seeing:
> 
> ![Screen Shot 2020-02-24 at 16 22 15](https://user-images.githubusercontent.com/973550/75164971-ef62fa80-5721-11ea-8c77-ff9984135354.png)

I think that is just from previous edits/reloads. Try opening the ng-run demo in a fresh incognito tab.

**Comment by pkozlowski-opensource** [comment]
> I think that is just from previous edits/reloads. Try opening the ng-run demo in a fresh incognito tab.

@Achilles1515 ha! you are right. Regarding `LComponentView_TabComponent` - my understanding is that it should be present in the dev mode only (`window.ngDevMode` being truthy). A project build with AOT in production mode should not have this problem.

**Comment by Achilles1515** [comment]
> @Achilles1515 ha! you are right. Regarding `LComponentView_TabComponent` - my understanding is that it should be present in the dev mode only (`window.ngDevMode` being truthy). A project build with AOT in production mode should not have this problem.

Cool. So I believe this issue can be closed.
Yeah...ng-run is great, but finicky. I also just recently realized that the `ng-run.preview.firebaseapp.com` heap dumps share memory across open tabs. So I really should have said "Try opening the ng-run demo in a fresh incognito **window**."

**Comment by pkozlowski-opensource** [comment]
> Cool. So I believe this issue can be closed.

I will keep it open for a bit - I would like to verify the `LComponentView_TabComponent` thing.

**Comment by JohnnyDevNull** [comment]
Actually we have this problem to?
Are there any news or plans to clear them in dev mode also?

It would be perfekt to just be able to debug memory snapshots like with ViewEngine. There the Components are getting destroyed as excepted, but now in Ivy we have a memory leak with that LComponentView thing.

First snapshot:

![image](https://user-images.githubusercontent.com/5759334/80078482-454ff500-854f-11ea-96db-c91ae361f3bb.png)

Second snapshot after several revistis of the component:

![image](https://user-images.githubusercontent.com/5759334/80078533-5567d480-854f-11ea-81d6-14e373fae868.png)

**Comment by pkozlowski-opensource** [comment]
> Just ignore it because it's DevMode should be never an acceptable answer!

@JohnnyDevNull it is _definitively_ not out our intention to ignore memory leaks just because those occur in dev mode. Agree with you on the principle! 

For the record, I can still reproduce this situation in the latest Angular running ivy using the following stackblitz: https://stackblitz.com/edit/angular-xzeiex?file=src/app/app.module.ts

Looking at the mem snapshot we can see that the culprit is `LVIEW_COMPONENT_CACHE` (https://github.com/angular/angular/blob/d1ea1f4c7f3358b730b0d94e65b00bc28cae279c/packages/core/src/render3/instructions/lview_debug.ts#L59-L61):

![Screen Shot 2020-05-29 at 12 30 46](https://user-images.githubusercontent.com/973550/83250339-43a9ca80-a1a8-11ea-9138-61414ba8eee4.png)

**Comment by JohnnyDevNull** [comment]
Hi there, we recently recognized that the `LCompontentView` thing now is also leakting in ViewEngine mode:

![image](https://user-images.githubusercontent.com/5759334/86107417-16f81400-bac2-11ea-92eb-29be697c0180.png)

... as a side notice: We are using a custom ReuseStrategy, there we are destroing the components as followed:

```ts
  public destroy(navPath: string): void {
    Array.from(this.storedRouteHandles.entries())
      .filter(([route]) => route.startsWith(navPath))
      .forEach(([route, handle]) => {
        ((handle as any).componentRef as ComponentRef<any>).destroy();
        this.storedRouteHandles.delete(route);
      });
  }
```

**Comment by JoostK** [comment]
I can no longer reproduce using the originally reported ng-run example, which I ported to 11.0.rc, 10.0.0, 9.0.0 and 9.1.0. Only 9.0.0 shows the issue, where `TabComponent` is retained. Looking at a memory dump reveals that #35156 has addressed the retainer path that was causing trouble here, such that the component instance can be cleaned up. That landed in 9.1.0, so this particular leak has been fixed a while ago.

As a note, the `LComponentView_TabComponent` that is retained through `LVIEW_COMPONENT_CACHE` is a red herring, that's only the empty named array that is present for easier debugging/memory inspections. The size of these objects is negligible.

I'll go ahead and close this particular instance as already fixed. If anyone is still seeing memory retention issues in Ivy, please open a new issue with reproduction for further analysis.
